### ISSUE TITLE
SP4 pen input latency

### ISSUE THREAD

**Issue by lviggiani** [linux-surface/linux-surface#1061]
I know this was discussed already in the past, but I cannot recall where. Sorry for posting this again.

On SP4, when I move the pen on the screen the cursor is about .5s behind pen movement. I know that this is done for smoothing pen movement and reduce jittering on drawn lines but I'd like to check if I can tweak settings or source code to try to reduce that latency and make the pen more usable while writing or drawing more precisely.
Which module / settings are involved in this?
Thanks!

**Comment by qzed** [comment]
@StollD do we do smoothing on the SP4? I thought that kind of stuff was SP7 and later only.

So depending on whether we do that or not I'm not sure there is an easy way to fix latency.

**Comment by StollD** [comment]
No, we don't do anything like that. On SP4, we receive pen coordinates directly from the hardware.

However, the hardware expects to be used with GuC submission. When data is available, it will increase a number called the doorbell. This used to trigger an interrupt in the GuC firmware. Since we cannot use GuC submission anymore, the kernel needs to poll this number to detect incoming data. I wouldn't be suprised if this introduces additional latency.

Another source of increased latency can also be the processing in userspace, instead of kernel -> GuC -> kernel -> userspace we have to do kernel -> userspace -> kernel -> userspace which is three times the amount of context switches.

Could you maybe upload a video of this? I would like to see how it looks on your device. On my SB2, there is some latency, but it is very much possible to write with the pen. But I dont do any kind of precision stuff, so I'd like to know if your device behaves differently, or if I can live with how it behaves.

**Comment by 1701** [comment]
same problem here with my sp4

**Comment by lviggiani** [comment]
Here we go!

https://user-images.githubusercontent.com/139040/221016895-1ecdaee0-6f63-423f-9d07-f8dcad149f84.mp4

**Comment by StollD** [comment]
Yeah, this looks a lot like what I get on SB2. So it is not a problem with SP4, but a problem with how we have to process the data from the hardware.

I think the latency can be reduced by increasing the polling frequency in the kernel, but not by much, and the CPU usage is already excessive with the current settings imho (up to 20% when you have the pen in proximity for a while).

**Comment by lviggiani** [comment]
Ok thanks for the explanation. It's a pity because the pen this way is barely usable for writing and drawing (especially in the long period). BTW I understand the limitation and I guess that re-enabling GuC submission is not a viable option.
Is the new driver for the SP7+ models performing better?

**Comment by danielzgtg** [comment]
I thought of a radical strategy. We can patch the CPUIdle driver. One of the CPUs can mwait whatever memory address that we are polling. Further logic is then added to handle this CPU being woken up for normal reasons. Tickless and deeper sleep are to be disabled to ensure the polling occurs at a fixed rate. Upon detecting a change, iptsd is scheduled. Putting the polling in the scheduler is as fast of a low latency that is possible to get the data to iptsd because it's the scheduler itself which executes before resuming every process.

**Comment by StollD** [comment]
I did some very basic testing (aka. running `top` and then moving the stylus on the screen like crazy), and it turns out that `usleep_range(1000, 5000)` actually has a similar CPU usage to `msleep(20)`, sometimes even a lower one.

And on my device, data is generated with 100Hz, so with a 1ms to 5ms polling interval we should be well above that. So any additional latency would come from having to do processing in userspace.

**Comment by StollD** [comment]
I have also done some tests with the old driver that parsed the stylus data in the kernelspace instead of passing them to iptsd. Maybe it is me, but I honestly cannot make out a difference between the kernelspace driver (with all sleeps in the polling loop removed, so 100% CPU usage), and the current driver with `usleep_range` and iptsd.

@lviggiani Could you try building and loading https://github.com/linux-surface/intel-precise-touch? And then see if that improves the behaviour?

I would also be interested if this is also how the stylus behaves on windows. Because either the latency comes from having to poll at all (which I kinda doubt), or from something else in userspace (e.g. libinput), or its simply how the hardware behaves.

**Comment by lviggiani** [comment]
> @lviggiani Could you try building and loading https://github.com/linux-surface/intel-precise-touch? And then see if that improves the behaviour?

Sure! I'll try tonight and let you kown

> I would also be interested if this is also how the stylus behaves on windows. Because either the latency comes from having to poll at all (which I kinda doubt), or from something else in userspace (e.g. libinput), or its simply how the hardware behaves.

I removed windows long time ago on that device but I remember that the pen was working very smooth. My wife was using it for writing and for drawing at professional level. It was not so smooth as the iPad pro, but close to it.

**Comment by PhilDevProg** [comment]
> @lviggiani Could you try building and loading https://github.com/linux-surface/intel-precise-touch? And then see if that improves the behaviour?

On my Go 3, the pen also has this latency so I tried this but also can't see any difference.

**Comment by StollD** [comment]
Because the Go line is not using IPTS or iptsd.

**Comment by PhilDevProg** [comment]
> Because the Go line is not using IPTS or iptsd.

Oh, I didn't know that. Sorry. Should I open a new issue regarding the latency (seems to be the same as in this issue) of the Go 3 or is it normal?

**Comment by StollD** [comment]
Since the touchscreen on the Go line is supported upstream and doesn't depend on a processing component like iptsd, there wouldn't be anything we could do. So opening an issue wouldn't really help. 

But it does confirm my suspicion that a large part of the latency comes in fact from userspace (or from parts of the kernel that arent drivers, like scheduling).

**Comment by danielzgtg** [comment]
If scheduling a problem we can boost the priority like PulseAudio does. It uses RealtimeKit to reduce preemption so that audio is not interrupted. We can have iptsd set its own priority to the same of that of PulseAudio, and look into adding RealtimeKit support to whatever drawing app you have trouble with.

**Comment by lviggiani** [comment]
> @lviggiani Could you try building and loading https://github.com/linux-surface/intel-precise-touch? And then see if that improves the behaviour?

Hi, sorry for being so late... I just tried but I see no difference. Here is what I did:

- cloned repo
- `make`
- `sudo rmmod ipts.ko`
- `sudo insmod src/ipts.ko`

Is it correct?

**Comment by StollD** [comment]
Yes, that should be correct.

I did some research: Microsoft claims, that the pen on the SP4 has a latency of 45 milliseconds, while on the SP5 and later it should have 21 milliseconds.

I tried to "scientifically" measure the latency, aka. by recording the screen with my phone camera and then inspecting the frames to see how much time passed between the pen hitting the display and the reaction on the screen. In my case, about 100 milliseconds pass between both. I am also using a third party stylus, no idea if that might affect the results.

Setting the nice value of all associated processes (iptsd and the kernel thread that does the polling) to -20 doesn't change anything about this.

You can actually find reports of the pen showing a similar behavior on Windows: https://www.reddit.com/r/Surface/comments/phbul9/surface_pen_has_a_ton_of_delay_please_help/. It also seems to be dependent on the app, I have seen people report that OneNote has no latency at all while apps like Krita or Photoshop have visible latency.

I will try to do some more experiments later, like measuring the latency with an unlocked kernel driver that polls all the time, and with the driver that does the data parsing in kernel space. But I am starting to believe that a high latency is normal for this hardware and that Microsoft uses prediction algorithms to reduce it (maybe through the Windows Ink APIs?). 

The only way to know for sure would be to reinstall windows and measure the latency there. Maybe someone who dual boots would be willing to run some tests?

**Comment by StollD** [comment]
Thats interesting!

I applied that variable but I don't really know how to describe the result. It feels a bit more jittery, but not really less laggy.

Also, fyi, my tests that I talked above were made with a version of iptsd-show that displays stylus events, so the latency of 100ms that I measured would be before libinput adds anything on top.

**Comment by innocentDE** [comment]
Is there any feasible solution now? Pen latency is still not as good as on Windows.

**Comment by StollD** [comment]
The only feasible solution is to do the same that Windows appears to be doing and implement input prediction in every application that needs it.

**Comment by rexendevar** [comment]
@0skxr How did you do this? I thought I followed all the steps on my surface book 1 but I'm not seeing any difference. Libinput says I've successfully set the quirk for every device in my list whose name starts with IPTS.

**Comment by rexendevar** [comment]
> I could offer to add a module parameter to the IPTS driver so you can increase the polling frequency in exchange for increased CPU usage while writing.

Did this ever get done? I managed to turn off the libinput smoothing but it's not enough to solve the problem.
